File: pkg/iban/iban.go
// Package iban generates and validates Irish (IE) IBANs using the ISO
// 7064 mod-97-10 check-digit algorithm. It knows nothing about any
// particular bank's account-numbering scheme — callers supply the BBAN
// parts (bank code, sort code, account number) and get back a
// checksum-valid IBAN, or validate an arbitrary input string.
package iban

import (
	"fmt"
	"math/big"
	"strings"
)

// CountryIE is the ISO 3166-1 alpha-2 country code this package generates
// IBANs for. Validate also accepts other countries' IBANs (the mod-97
// checksum is country-agnostic per ISO 7064) but only enforces IE's fixed
// field layout.
const CountryIE = "IE"

// IELength is the fixed length of a normalized IE IBAN. Exported because
// Validate only enforces this shape for country code "IE" — for any other
// country it just checks the generic checksum, with no minimum length
// beyond 4 — so a caller that needs to safely slice into the fixed IE BBAN
// layout (bank code, sort code, account number) must check both the
// country code AND this length first, not assume Validate already implies
// it for arbitrary input.
const IELength = 2 + 2 + bankCodeLength + sortCodeLength + bbanAccountNumberLength // 22

const (
	bankCodeLength          = 4 // letters
	sortCodeLength          = 6 // digits
	bbanAccountNumberLength = 8 // digits
	mod97Modulus            = 97
)

var mod97 = big.NewInt(mod97Modulus)

// Generate builds a checksum-valid IE IBAN from BBAN parts: a 4-letter
// bank/institution code, a 6-digit sort code, and an 8-digit account
// number. bankCode is upper-cased before use; sortCode and accountNumber
// must already be digit strings of the exact required length.
func Generate(bankCode, sortCode, accountNumber string) (string, error) {
	bankCode = strings.ToUpper(bankCode)
	if len(bankCode) != bankCodeLength || !isAllLetters(bankCode) {
		return "", fmt.Errorf("iban: bank code must be %d letters, got %q", bankCodeLength, bankCode)
	}
	if len(sortCode) != sortCodeLength || !isAllDigits(sortCode) {
		return "", fmt.Errorf("iban: sort code must be %d digits, got %q", sortCodeLength, sortCode)
	}
	if len(accountNumber) != bbanAccountNumberLength || !isAllDigits(accountNumber) {
		return "", fmt.Errorf("iban: account number must be %d digits, got %q", bbanAccountNumberLength, accountNumber)
	}

	bban := bankCode + sortCode + accountNumber

	// ISO 7064 mod-97-10: compute the check digits by first treating them
	// as "00", i.e. rearranging BBAN + country + "00" (the "move the first
	// four characters to the end" step applied to a not-yet-existing IBAN
	// whose check digits are provisionally zero).
	remainder, err := mod97Remainder(bban + CountryIE + "00")
	if err != nil {
		return "", err
	}
	checkDigits := 98 - remainder
	return fmt.Sprintf("%s%02d%s", CountryIE, checkDigits, bban), nil
}

// Validate normalizes s (strips whitespace, upper-cases — see Normalize)
// and reports whether it is a structurally valid, checksum-correct IBAN.
// The mod-97 checksum check applies to any country's IBAN; the fixed
// 4-letter/6-digit/8-digit field layout is only enforced for IE, since
// that's the only country format this package has authoritative lengths
// for.
func Validate(s string) error {
	n := Normalize(s)

	if len(n) < 4 {
		return fmt.Errorf("iban: %q is too short to be an IBAN", s)
	}
	country := n[0:2]
	checkDigits := n[2:4]
	bban := n[4:]

	if !isAllLetters(country) {
		return fmt.Errorf("iban: %q has an invalid country code %q", s, country)
	}
	if !isAllDigits(checkDigits) {
		return fmt.Errorf("iban: %q has non-numeric check digits %q", s, checkDigits)
	}
	if !isAllLettersOrDigits(bban) {
		return fmt.Errorf("iban: %q contains characters other than letters and digits", s)
	}

	if country == CountryIE {
		if len(n) != IELength {
			return fmt.Errorf("iban: IE IBANs must be %d characters, %q is %d", IELength, s, len(n))
		}
		bankCode := bban[0:bankCodeLength]
		sortCode := bban[bankCodeLength : bankCodeLength+sortCodeLength]
		acctNum := bban[bankCodeLength+sortCodeLength:]
		if !isAllLetters(bankCode) {
			return fmt.Errorf("iban: %q has an invalid bank code %q", s, bankCode)
		}
		if !isAllDigits(sortCode) {
			return fmt.Errorf("iban: %q has an invalid sort code %q", s, sortCode)
		}
		if !isAllDigits(acctNum) {
			return fmt.Errorf("iban: %q has an invalid account number %q", s, acctNum)
		}
	}

	// Validation rearranges the SAME way generation does, but with the
	// IBAN's actual check digits rather than a "00" placeholder: a
	// correct IBAN's rearranged numeric form is congruent to 1 mod 97.
	remainder, err := mod97Remainder(bban + country + checkDigits)
	if err != nil {
		return fmt.Errorf("iban: %q: %w", s, err)
	}
	if remainder != 1 {
		return fmt.Errorf("iban: %q fails the mod-97 checksum", s)
	}
	return nil
}

// Normalize returns the canonical storage form: no whitespace, upper case.
func Normalize(s string) string {
	return strings.ToUpper(strings.ReplaceAll(s, " ", ""))
}

// Format groups the normalized form into 4-character blocks separated by
// spaces, for display — e.g. "IE29 AIBK 9311 5212 3456 78".
func Format(s string) string {
	n := Normalize(s)
	var b strings.Builder
	for i := 0; i < len(n); i += 4 {
		if i > 0 {
			b.WriteByte(' ')
		}
		end := i + 4
		if end > len(n) {
			end = len(n)
		}
		b.WriteString(n[i:end])
	}
	return b.String()
}

// mod97Remainder converts s (a rearranged IBAN candidate: BBAN + country +
// check digits, in that order) to its ISO 7064 numeric form — letters
// become two-digit numbers (A=10 .. Z=35), digits pass through — and
// returns that (very large) number mod 97. math/big is required: a 22-char
// IE IBAN's numeric form runs to ~28 digits, well past int64 range.
func mod97Remainder(s string) (int64, error) {
	digits, err := lettersToDigits(s)
	if err != nil {
		return 0, err
	}
	n, ok := new(big.Int).SetString(digits, 10)
	if !ok {
		return 0, fmt.Errorf("iban: could not parse %q as a number", digits)
	}
	return new(big.Int).Mod(n, mod97).Int64(), nil
}

// lettersToDigits implements ISO 7064's letter substitution: each letter
// A-Z becomes its two-digit position (A=10 ... Z=35), each digit passes
// through unchanged.
func lettersToDigits(s string) (string, error) {
	var b strings.Builder
	b.Grow(len(s) * 2)
	for _, r := range s {
		switch {
		case r >= '0' && r <= '9':
			b.WriteRune(r)
		case r >= 'A' && r <= 'Z':
			fmt.Fprintf(&b, "%d", r-'A'+10)
		default:
			return "", fmt.Errorf("iban: invalid character %q", r)
		}
	}
	return b.String(), nil
}

func isAllLetters(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if r < 'A' || r > 'Z' {
			return false
		}
	}
	return true
}

func isAllDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return true
}

func isAllLettersOrDigits(s string) bool {
	for _, r := range s {
		isLetter := r >= 'A' && r <= 'Z'
		isDigit := r >= '0' && r <= '9'
		if !isLetter && !isDigit {
			return false
		}
	}
	return true
}

File: frontend/src/features/transfers/iban.ts
// Client-side mirror of pkg/iban/iban.go's Validate/Normalize/Format —
// same ISO 7064 mod-97-10 check-digit algorithm, kept in lockstep so the
// form can reject a malformed IBAN before it ever reaches the backend.
// Uses native BigInt instead of math/big; JS BigInt has no size limit, so
// the ~28-digit numeric form of an IE IBAN is exact here too.

export const COUNTRY_IE = 'IE'

const BANK_CODE_LENGTH = 4 // letters
const SORT_CODE_LENGTH = 6 // digits
const ACCOUNT_NUMBER_LENGTH = 8 // digits
export const IE_LENGTH = 2 + 2 + BANK_CODE_LENGTH + SORT_CODE_LENGTH + ACCOUNT_NUMBER_LENGTH // 22

const MOD97 = 97n

// Normalize returns the canonical storage form: no whitespace, upper case.
export function normalizeIban(s: string): string {
  return s.toUpperCase().replace(/\s+/g, '')
}

// Format groups the normalized form into 4-character blocks separated by
// spaces, for display — e.g. "IE29 AIBK 9311 5212 3456 78". Safe to call
// on partial input while the user is still typing.
export function formatIban(s: string): string {
  const n = normalizeIban(s)
  const blocks: string[] = []
  for (let i = 0; i < n.length; i += 4) {
    blocks.push(n.slice(i, i + 4))
  }
  return blocks.join(' ')
}

function isAllLetters(s: string): boolean {
  return s.length > 0 && /^[A-Z]+$/.test(s)
}

function isAllDigits(s: string): boolean {
  return s.length > 0 && /^[0-9]+$/.test(s)
}

function isAllLettersOrDigits(s: string): boolean {
  return /^[A-Z0-9]*$/.test(s)
}

// lettersToDigits implements ISO 7064's letter substitution: each letter
// A-Z becomes its two-digit position (A=10 ... Z=35), each digit passes
// through unchanged. Returns null on any other character.
function lettersToDigits(s: string): string | null {
  let out = ''
  for (const ch of s) {
    if (ch >= '0' && ch <= '9') {
      out += ch
    } else if (ch >= 'A' && ch <= 'Z') {
      out += String(ch.charCodeAt(0) - 'A'.charCodeAt(0) + 10)
    } else {
      return null
    }
  }
  return out
}

function mod97Remainder(s: string): bigint | null {
  const digits = lettersToDigits(s)
  if (digits === null) return null
  return BigInt(digits) % MOD97
}

// validateIban mirrors pkg/iban/iban.go's Validate: it reports why `s`
// isn't a valid IBAN, or null if it is. The mod-97 checksum applies to any
// country's IBAN; the fixed 4-letter/6-digit/8-digit field layout is only
// enforced for IE, matching the backend (the only country this package has
// an authoritative layout for).
export function validateIban(s: string): string | null {
  const n = normalizeIban(s)

  if (n.length < 4) return 'IBAN is too short'

  const country = n.slice(0, 2)
  const checkDigits = n.slice(2, 4)
  const bban = n.slice(4)

  if (!isAllLetters(country)) return 'Invalid country code'
  if (!isAllDigits(checkDigits)) return 'Invalid check digits'
  if (!isAllLettersOrDigits(bban)) return 'IBAN contains invalid characters'

  if (country === COUNTRY_IE) {
    if (n.length !== IE_LENGTH) return `An Irish IBAN must be ${IE_LENGTH} characters`
    const bankCode = bban.slice(0, BANK_CODE_LENGTH)
    const sortCode = bban.slice(BANK_CODE_LENGTH, BANK_CODE_LENGTH + SORT_CODE_LENGTH)
    const acctNum = bban.slice(BANK_CODE_LENGTH + SORT_CODE_LENGTH)
    if (!isAllLetters(bankCode)) return 'Invalid bank code'
    if (!isAllDigits(sortCode)) return 'Invalid sort code'
    if (!isAllDigits(acctNum)) return 'Invalid account number'
  }

  const remainder = mod97Remainder(bban + country + checkDigits)
  if (remainder === null) return 'IBAN contains invalid characters'
  if (remainder !== 1n) return 'Invalid IBAN check digits'

  return null
}

export function isValidIban(s: string): boolean {
  return validateIban(s) === null
}
